numba/np/random/generator_methods.py

```diff
@@ -166,6 +167,59 @@ def impl(inst, low, high, size=None,
         return impl
 
 
+# The following `shuffle` implementation is a direct translation from:
+# https://github.com/numpy/numpy/blob/95e3e7f445407e4f355b23d6a9991d8774f0eb0c/numpy/random/_generator.pyx#L4578
+
+# Overload the Generator().shuffle()
+@overload_method(types.NumPyRandomGeneratorType, 'shuffle')
+def NumPyRandomGeneratorType_shuffle(inst, x, axis=0):
+    check_types(x, [types.Array], 'x')
+    check_types(axis, [int, types.Integer], 'axis')
+    axis_impl = tuple([slice(None, None)] * x.ndim)
+
+    def impl(inst, x, axis=0):
+        if axis > x.ndim - 1:
+            raise IndexError("Axis is out of bounds for the given array")
+        for i in range(x.shape[axis] - 1, 0, -1):
+            j = types.intp(random_methods.random_interval(inst.bit_generator,
+                                                          i))
+            if i == j:
+                continue
+            axis_impl1 = tuple_setitem(axis_impl, axis, slice(i, i + 1))
+            axis_impl2 = tuple_setitem(axis_impl, axis, slice(j, j + 1))
+
+            buffer = np.copy(x[axis_impl2])
+            x[axis_impl2] = np.copy(x[axis_impl1])
```

Comment:
I think this might end up being quite slow because it's using a multiple new buffers for every iteration of this loop? Following OOB discussion, perhaps a rewrite to implement this closer to how NumPy implements this loop would be a good idea. i.e. (based on the same code in NumPy's `95e3e7f` as referenced):
```python
    def impl(inst, x, axis=0):
        if axis > x.ndim - 1:
            raise IndexError("Axis is out of bounds for the given array")

        z = np.swapaxes(x, 0, axis)
        buf = np.empty_like(z[0, ...])

        for i in range(len(z) - 1, 0, -1):
            j = types.intp(random_methods.random_interval(inst.bit_generator,
                                                          i))
            if i == j:
                continue
            buf[...] = z[j, ...]
            z[j, ...] = z[i, ...]
            z[i, ...] = buf

    return impl
```

Reply:
At first glance it seems like `swapaxes` approach doesn't copy anything at all. 

The following implementation: 
```python
@overload_method(types.NumPyRandomGeneratorType, 'shuffle')
def NumPyRandomGeneratorType_shuffle(inst, x, axis=0):
    check_types(x, [types.Array], 'x')
    check_types(axis, [int, types.Integer], 'axis')
    axis_impl = tuple([slice(None, None)] * x.ndim)

    def impl(inst, x, axis=0):
        x = np.swapaxes(x, 0, axis)
        buf = np.empty_like(x[0, ...])
        for i in range(x.shape[axis] - 1, 0, -1):
            j = types.intp(random_methods.random_interval(inst.bit_generator,
                                                          i))
            if i == j:
                # i == j is not needed and memcpy is undefined.
                continue
            buf[...] = np.copy(x[j])
            x[j] = np.copy(x[i])
            x[i] = buf[...]
```
 does nothing to the array.


I have one more implementation that works:
```python
@overload_method(types.NumPyRandomGeneratorType, 'shuffle')
def NumPyRandomGeneratorType_shuffle(inst, x, axis=0):
    check_types(x, [types.Array], 'x')
    check_types(axis, [int, types.Integer], 'axis')
    axis_impl = tuple([slice(None, None)] * x.ndim)
    _shape_tuple = tuple([0]*(x.ndim-1))

    def impl(inst, x, axis=0):

        if axis > x.ndim - 1:
            raise IndexError("Axis is out of bounds for the given array")
        idx = 0
        shape_tuple = _shape_tuple

        for i in range(x.ndim):
            if i==axis:
                continue
            shape_tuple = tuple_setitem(shape_tuple, idx, x.shape[i])
            idx = idx+1

        buffer = np.empty(shape_tuple)

        for i in range(x.shape[axis] - 1, 0, -1):
            j = types.intp(random_methods.random_interval(inst.bit_generator,
                                                          i))
            if i == j:
                continue
            axis_impl1 = tuple_setitem(axis_impl, axis, slice(i, i + 1))
            axis_impl2 = tuple_setitem(axis_impl, axis, slice(j, j + 1))

            buffer = np.copy(x[axis_impl2])
            x[axis_impl2] = np.copy(x[axis_impl1])
            x[axis_impl1] = buffer

    return impl
```
but I'm not sure if this prevents the creation of new buffers at every iteration ?

Another implementation but memory inefficient (create a buffer of size of array itself):
```python
@overload_method(types.NumPyRandomGeneratorType, 'shuffle')
def NumPyRandomGeneratorType_shuffle(inst, x, axis=0):
    check_types(x, [types.Array], 'x')
    check_types(axis, [int, types.Integer], 'axis')
    axis_impl = tuple([slice(None, None)] * x.ndim)

    def impl(inst, x, axis=0):
        if axis > x.ndim - 1:
            raise IndexError("Axis is out of bounds for the given array")
        buffer = np.empty_like(x)
        for i in range(x.shape[axis] - 1, 0, -1):
            j = types.intp(random_methods.random_interval(inst.bit_generator,
                                                          i))
            if i == j:
                continue
            axis_impl1 = tuple_setitem(axis_impl, axis, slice(i, i + 1))
            axis_impl2 = tuple_setitem(axis_impl, axis, slice(j, j + 1))

            buffer[axis_impl2] = np.copy(x[axis_impl2])
            x[axis_impl2] = np.copy(x[axis_impl1])
            x[axis_impl1] = np.copy(buffer[axis_impl2])
```